File: LeetCode/1414. Find the Minimum Number of Fibonacci Numbers Whose Sum Is K/solution.py
class Solution(object):
    def findMinFibonacciNumbers(self, k):
        fib = [1, 1]
        while fib[-1] < k:
            fib.append(fib[-1] + fib[-2])

        ans = 0
        while k:
            x = fib.pop()
            if k >= x:
                k -= x
                ans += 1
        return ans

Zeckendorf representation (ZR) of K (a representation of K as a sum of non-adjacent Fibonacci values)

https://discordapp.com/channels/612060087900438538/612587060099940372/701101422241906729

# class Solution:
#     def findMinFibonacciNumbers(self, k: int) -> int:
#         fib = [1,1]
#         while fib[-1] < k:
#             fib.append(fib[-1]+fib[-2])
#         # print(fib)
#         step = 0
#         while k > 0:
#             idx = bisect.bisect_right(fib,k)-1
#             k -= fib[idx]
#             step += 1
#         return step

File: LeetCode/1414. Find the Minimum Number of Fibonacci Numbers Whose Sum Is K/solution.cpp
const int maxn = 45;
int fib[maxn];
bool initialized = false;

class Solution {
public:
    void init(){
        fib[0] = fib[1] = 1;
        for (int j = 2; j < maxn; j++){
            fib[j] = fib[j-1] + fib[j-2];
            // cout << fib[j] << " ";
        }
        // cout << fib[maxn-1] << endl;
    }
    int findMinFibonacciNumbers(int k) {
        if (!initialized)
            init();
        int cnt = 0;
        for (int j = maxn-1; j >= 0 && k; j--){
            if (fib[j] <= k){
                cnt++;
                k -= fib[j];
            }
        }
        return cnt;
    }
};